patches/api/0387-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,286 @@
+@@ -13,6 +13,7 @@ import org.bukkit.entity.Player;
+ import org.bukkit.event.Cancellable;
+ import org.bukkit.event.HandlerList;
+ import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
+ 
+ /**
+  * Stores data for damage events
+@@ -25,13 +26,19 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
+     private final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions;
+     private final Map<DamageModifier, Double> originals;
+     private boolean cancelled;
+-    private final DamageCause cause;
++    private final com.destroystokyo.paper.entity.damage.DamageOrigin cause; // Paper
+ 
++    @Deprecated // Paper
+     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage) {
+         this(damagee, cause, new EnumMap<DamageModifier, Double>(ImmutableMap.of(DamageModifier.BASE, damage)), new EnumMap<DamageModifier, Function<? super Double, Double>>(ImmutableMap.of(DamageModifier.BASE, ZERO)));
+     }
+ 
++    @Deprecated // Paper
+     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
++        // Paper start - Damage source wrapper
++        this(damagee, (com.destroystokyo.paper.entity.damage.DamageOrigin) null, modifiers, modifierFunctions);
++    }
++    public EntityDamageEvent(@NotNull final Entity damagee, final com.destroystokyo.paper.entity.damage.DamageOrigin cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
+         super(damagee);
+         Preconditions.checkArgument(modifiers.containsKey(DamageModifier.BASE), "BASE DamageModifier missing");
+         Preconditions.checkArgument(!modifiers.containsKey(null), "Cannot have null DamageModifier");
+@@ -43,6 +50,7 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
+         this.modifiers = modifiers;
+         this.modifierFunctions = modifierFunctions;
+     }
++    // Paper end
+ 
+     @Override
+     public boolean isCancelled() {
+@@ -182,13 +190,15 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
+         setDamage(DamageModifier.BASE, damage);
+     }
+ 
++    // Paper start - damage source wrapper
+     /**
+      * Gets the cause of the damage.
+      *
+-     * @return A DamageCause value detailing the cause of the damage.
++     * @return A DamageOrigin value detailing the cause of the damage.
+      */
+-    @NotNull
+-    public DamageCause getCause() {
++    @Nullable // null for deprecated constructor
++    public com.destroystokyo.paper.entity.damage.DamageOrigin getCause() {
++        // Paper end - damage source wrapper
```

Comment:
You can't just remove the old damagcause like this, you will need to have both here.